### PR TITLE
[Core] Add Command to migrate content between storages

### PR DESCRIPTION
**Pull request by dvesh3**
## Changes in this pull request  
Follow up to https://github.com/pimcore/pimcore/pull/8562

## Additional info  
Added `MigrateStorageCommand` for migrating existing content between storages e.g. local filesystem to AWS S3.

Steps:
1. Add config:
```yaml
flysystem:
    storages:
        pimcore.asset.storage.source:
            adapter: 'local'
            visibility: public
            options:
                directory: '%kernel.project_dir%/public/var/assets'
        pimcore.asset.storage.target:
            adapter: 'aws'
            visibility: public
            options:
                client: 'assets_s3'
                bucket: 'bucket-name'
                prefix: asset
```
2. Run command: `bin/console pimcore:migrate:storage asset`

### CODE REVIEW

File: doc/Development_Documentation/23_Installation_and_Upgrade/03_System_Setup_and_Hosting/03_File_Storage_Setup.md [hunk 1]
```diff
@@ -111,4 +159,63 @@ flysystem:
...
+```yaml
+flysystem:
+    storages:
+        pimcore.asset_source.storage:
+            adapter: 'local'
+            visibility: public
+            options:
+                directory: '%kernel.project_dir%/public/var/assets'
+
+        pimcore.asset_destination.storage:
```
Comment (brusch) [inline doc/Development_Documentation/23_Installation_and_Upgrade/03_System_Setup_and_Hosting/03_File_Storage_Setup.md]:
```suggestion
        pimcore.asset.storage.destination:
```
@dvesh3 What do you think about this naming convention?

**Review by brusch — CHANGES_REQUESTED** [review]

File: doc/Development_Documentation/23_Installation_and_Upgrade/03_System_Setup_and_Hosting/03_File_Storage_Setup.md (hunk 1 above)
Comment (dvesh3) [reply doc/Development_Documentation/23_Installation_and_Upgrade/03_System_Setup_and_Hosting/03_File_Storage_Setup.md]: @brusch Right now we are using `Storage::get()` which uses format `pimcore.%s.storage` to get the service https://github.com/pimcore/pimcore/blob/10.x/lib/Tool/Storage.php#L37.

Comment (dvesh3) [reply doc/Development_Documentation/23_Installation_and_Upgrade/03_System_Setup_and_Hosting/03_File_Storage_Setup.md]: maybe `pimcore.asset.destination.storage`?

Comment (brusch) [reply doc/Development_Documentation/23_Installation_and_Upgrade/03_System_Setup_and_Hosting/03_File_Storage_Setup.md]: Aaaah, I see. But we can just use the locator directly in the command, right? In my opinion a suffix would be just easier and looks a bit nicer.

Comment (dvesh3) [reply doc/Development_Documentation/23_Installation_and_Upgrade/03_System_Setup_and_Hosting/03_File_Storage_Setup.md]: Updated the command to use locator directly and use suffix for storage names

**Comment by brusch** [comment]
A possible improvement could be using the parallelization trait to speed things up - maybe someone from the community wants to pick up here? 😊

**Comment by solverat** [comment]
@dvesh3 What about migrating an existing s3 bucket with a new pimcore installation? I could do a simple `$storage->listContents('/', true)` to create  assets locally but in fact i only need the db-references.  So I've created a simple adapter to initial sync s3 data with the local db. 

```yaml
flysystem:
    storages:
        pimcore_local:
            adapter: 'App\Flysystem\PimcoreLocalAdapter'

        pimcore.asset.storage.source:
            adapter: 'aws'
            visibility: public
            options:
                client: 'assets_s3'
                bucket: 'MY_BUCKET'
                prefix: 'PREFIX'

        pimcore.asset.storage.target:
            adapter: 'pimcore_local'
            visibility: public

services:
    pimcore_local:
        autowire: true
        class: 'App\Flysystem\PimcoreLocalAdapter'
```

```php
<?php

namespace App\Flysystem;

use League\Flysystem\Config;
use League\Flysystem\FileAttributes;
use League\Flysystem\FilesystemAdapter;
use Pimcore\Db\Connection;
use Pimcore\Model\Asset;
use Pimcore\Model\Asset\Service;

class PimcoreLocalAdapter implements FilesystemAdapter
{
    protected Connection $connection;

    public function __construct(Connection $connection)
    {
        $this->connection = $connection;
    }

    public function writeStream(string $path, $contents, Config $config): void
    {
        if ($path === '') {
            return;
        }

        $dirName = dirname($path);
        $fileName = basename($path);

        $assetParent = Service::createFolderByPath($dirName);

        $fileInfo = new \finfo(FILEINFO_MIME_TYPE);
        $mimeType = $fileInfo->buffer(stream_get_contents($contents));

        $type = Asset::getTypeFromMimeMapping($mimeType, $path);

        $asset = new Asset();
        $asset->setType($type);
        $asset->setParent($assetParent);
        $asset->setKey($fileName);
        $asset->setDataChanged(false);

        $asset->save();
    }

    public function fileExists(string $path): bool
    {
        $dirName = dirname($path);
        $fileName = basename($path);

        return $this->connection->fetchRow(
                'SELECT id FROM assets WHERE path = :path AND `filename` = :key',
                [
                    'path' => sprintf('/%s/', $dirName),
                    'key'  => $fileName
                ]
            ) !== false;
    }

    public function write(string $path, string $contents, Config $config): void
    {
        throw new \Exception('Not implemented');
    }

    public function read(string $path): string
    {
        throw new \Exception('Not implemented');
    }

    public function readStream(string $path)
    {
        throw new \Exception('Not implemented');
    }

    public function delete(string $path): void
    {
        throw new \Exception('Not implemented');
    }

    public function deleteDirectory(string $path): void
    {
        throw new \Exception('Not implemented');
    }

    public function createDirectory(string $path, Config $config): void
    {
        throw new \Exception('Not implemented');
    }

    public function setVisibility(string $path, string $visibility): void
    {
        throw new \Exception('Not implemented');
    }

    public function visibility(string $path): FileAttributes
    {
        throw new \Exception('Not implemented');
    }

    public function mimeType(string $path): FileAttributes
    {
        throw new \Exception('Not implemented');
    }

    public function lastModified(string $path): FileAttributes
    {
        throw new \Exception('Not implemented');
    }

    public function fileSize(string $path): FileAttributes
    {
        throw new \Exception('Not implemented');
    }

    public function listContents(string $path, bool $deep): iterable
    {
        throw new \Exception('Not implemented');
    }

    public function move(string $source, string $destination, Config $config): void
    {
        throw new \Exception('Not implemented');
    }

    public function copy(string $source, string $destination, Config $config): void
    {
        throw new \Exception('Not implemented');
    }
}
```

Hope this helps someone out there!

**Comment by dvesh3** [comment]
@solverat thanks for sharing it 👍 maybe a good idea to post it on https://github.com/pimcore/pimcore/discussions/categories/ideas

**Comment by solverat** [comment]
good one. done & thanks!

**Pull request merged**